File: leetcode/go/2.add-two-numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	}
	if l2 == nil {
		return l1
	}
	var ret, tailNode *ListNode
	ret = nil
	l1Node := l1
	l2Node := l2
	addOne := 0
	firstVal := 0
	secondVal := 0
	for l1Node != nil || l2Node != nil {
		if l1Node != nil {
			firstVal = l1Node.Val
		} else {
			firstVal = 0
		}

		if l2Node != nil {
			secondVal = l2Node.Val
		} else {
			secondVal = 0
		}
		tmp := firstVal + secondVal + addOne
		if tmp > 9 {
			addOne = 1
		} else {
			addOne = 0
		}
		tmp = tmp % 10
		if ret == nil {
			ret = &ListNode{
				Val: tmp,
			}
			tailNode = ret
		} else {
			tailNode.Next = &ListNode{
				Val: tmp,
			}
			tailNode = tailNode.Next
		}
		if l1Node != nil {
			l1Node = l1Node.Next
		}
		if l2Node != nil {
			l2Node = l2Node.Next
		}

	}
	if addOne == 1 {
		tailNode.Next = &ListNode{
			Val: 1,
		}
	}
	return ret
}

// @lc code=end

File: leetcode/cpp/2.add-two-numbers.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include <iostream>
using namespace std;
// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };

class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        if (l1 == nullptr)
            return l2;
        if (l2 == nullptr)
            return l1;
        int tmp = 0;
        bool biggerThan9 = false;
        ListNode *l1Node = l1;
        ListNode *l2Node = l2;
        ListNode *remainNode = nullptr;
        ListNode *ret = nullptr;
        ListNode *retNode = ret;
        while (l1Node != nullptr && l2Node != nullptr)
        {
            tmp = l1Node->val + l2Node->val + (biggerThan9 ? 1 : 0);
            if (tmp > 9)
            {
                biggerThan9 = true;
                tmp %= 10;
            }
            else
            {
                biggerThan9 = false;
            }

            if (ret)
            {
                retNode->next = new ListNode(tmp);
                retNode = retNode->next;
            }
            else
            {
                ret = new ListNode(tmp);
                retNode = ret;
            }
            l1Node = l1Node->next;
            l2Node = l2Node->next;
        }
        remainNode = (l1Node == nullptr ? l2Node : l1Node);
        while (remainNode != nullptr)
        {
            tmp = (biggerThan9 ? 1 : 0) + remainNode->val;
            if (tmp > 9)
            {
                biggerThan9 = true;
                tmp %= 10;
            }
            else
            {
                biggerThan9 = false;
            }

            retNode->next = new ListNode(tmp);
            retNode = retNode->next;
            remainNode = remainNode->next;
        }
        if (biggerThan9)
        {
            retNode->next = new ListNode(1);
        }
        return ret;
    }
};
// @lc code=end
